apiserver/apiserver_test.go

```diff
@@ -0,0 +1,332 @@
+///////////////////////////////////////////////////////
+
+func TestNotFoundDeleteConfig(t *testing.T) {
+	mockReg := mock.MakeRegistry()
+	api := makeAPIServer(mockReg)
+	url := "/test/config/route-rule/namespace/name"
+	status, body := makeAPIRequest(api, "DELETE", url, validRouteRuleJSON, t)
+
+	compareStatus(status, http.StatusNotFound, t)
+	compareResponseText(string(body), notFoundResponse, t)
+
+}
+
+func TestInvalidConfigTypeDeleteConfig(t *testing.T) {
+	mockReg := mock.MakeRegistry()
+	api := makeAPIServer(mockReg)
+	url := "/test/config/not-a-route-rule/namespace/name"
+	status, _ := makeAPIRequest(api, "DELETE", url, validRouteRuleJSON, t)
+
+	compareStatus(status, http.StatusBadRequest, t)
+}
+
+///////////////////////////////////////////////////////
+///////////////// LIST CONFIG ERRORS //////////////////
+///////////////////////////////////////////////////////
+
+func TestInvalidConfigTypeWithNamespaceListConfig(t *testing.T) {
+	mockReg := mock.MakeRegistry()
+	api := makeAPIServer(mockReg)
+	url := "/test/config/not-a-route-rule/namespace"
+	status, _ := makeAPIRequest(api, "GET", url, nil, t)
+
+	compareStatus(status, http.StatusBadRequest, t)
+}
+
+func TestInvalidConfigTypeWithoutNamespaceListConfig(t *testing.T) {
+	mockReg := mock.MakeRegistry()
+	api := makeAPIServer(mockReg)
+	url := "/test/config/not-a-route-rule"
+	status, _ := makeAPIRequest(api, "GET", url, nil, t)
+
+	compareStatus(status, http.StatusBadRequest, t)
+}
+
+///////////////////////////////////////////////////////
+////////////////// HELPER FUNCTIONS ///////////////////
+///////////////////////////////////////////////////////
+
+func compareListCount(body []byte, expected int, t *testing.T) {
+	configSlice := []Config{}
+	if err := json.Unmarshal(body, &configSlice); err != nil {
+		t.Errorf("unexpected error: %v", err)
+	}
+	if len(configSlice) != expected {
+		t.Errorf("expected %v elements back but got %v", expected, len(configSlice))
+	}
+}
+
+func compareStatus(received, expected int, t *testing.T) {
+	if received != expected {
+		t.Errorf("Expected status code: %d, received: %d", expected, received)
+	}
+}
+
+func compareResponseText(received, expected string, t *testing.T) {
+	if strings.Compare(received, expected) != 0 {
+		t.Errorf("expected response body to be %v, but received: %v", expected, received)
+	}
+}
+
+func compareResponseJSON(body []byte, file string, t *testing.T) {
+	err := ioutil.WriteFile(file, body, 0644)
+	if err != nil {
+		t.Fatalf(err.Error())
+	}
+	compareJSON(file, t)
+}
+
+func compareStoredConfig(mockReg *model.IstioRegistry, key model.Key, present bool, t *testing.T) {
+	_, ok := mockReg.Get(key)
+	if !ok && present {
+		t.Errorf("Expected config wasn't present in the registry for key: %+v", key)
+	} else if ok && !present {
+		t.Errorf("Unexpected config was present in the registry for key: %+v", key)
+	}
+	// To Do: compare protos
+}
+
+func compareJSON(jsonFile string, t *testing.T) {
```

Comment:
please use `test/util/diff.go`. I added a flag to regenerate testdata files that should work across all usages of that diff library.